Modifying the `shard_num` read for templates to indicate that `data_path` flag needs to be passed as an integer

theakshaypant opened the pull request:
**Description**

This is a documentation patch.
Currently, docs seem to show `data_path` to be an optional flag. But for the specific example, it is mandatory. This PR makes the mention explicit.

In addition, a guardrail to expect `data_path` as a shard number is added in all affected tutorials.

**PR Changes**
- Updated the docs to represent that data_path is not an optional parameter for the tutorial. 
- Adds a detailed error message if the shard_num is not an integer value. 
- This change is made in the following templates:
  - keras_cnn_mnist
  - keras_cnn_with_compression
  - torch_cnn_histology
  - torch_cnn_histology_gramine_ready
  - torch_cnn_mnist
  - torch_cnn_mnist_eden_compression
  - torch_cnn_mnist_fed_eval
  - torch_cnn_mnist_straggler_check
  - torch_unet_kvasir
  - torch_unet_kvasir_gramine_ready

theakshaypant committed:
fix(workspaces): reading shard number from default path

Signed-off-by: Pant, Akshay <akshay.akshaypant@gmail.com>

MasterSkepticista reviewed (changes requested):
I think extracting first integer from the string is a dangerous assumption. From my understanding, those data paths (e.g. `data/c1` or `data/collab2`) are not actual locations on disk (beyond the fact that user may also provide `data/100shards/shard0`, wherein this regex will fail).

In this case, try casting `data_path` to `int` or raise a `ValueError` that asks user to pass shard ID in the `-d <shard_num>` format, since data path does not mean anything in this example.

theakshaypant committed:
fix(workspaces): added valueerror handling where data path represents shard num

Signed-off-by: Pant, Akshay <akshay.akshaypant@gmail.com>

theakshaypant committed:
fix(docs): removed optional from collaborator create command as omitting it leads to an error when starting the collaborator for some templates

Signed-off-by: Pant, Akshay <akshay.akshaypant@gmail.com>

theakshaypant commented:
> I think extracting first integer from the string is a dangerous assumption. From my understanding, those data paths (e.g. `data/c1` or `data/collab2`) are not actual locations on disk (beyond the fact that user may also provide `data/100shards/shard0`, wherein this regex will fail).
> 
> In this case, try casting `data_path` to `int` or raise a `ValueError` that asks user to pass shard ID in the `-d <shard_num>` format, since data path does not mean anything in this example.

- Added exception handling when reading shard number from data path in 57809a692f41bd3c78c5eff7d808231254101092. 
- Removed the optional tag with data path in the docs in 70fe34afac99ca3e6035a2e1f433bafd81def28d.

MasterSkepticista commented on openfl-workspace/keras_cnn_mnist/src/tfmnist_inmemory.py:
```diff
@@ -27,9 +30,13 @@ def __init__(self, data_path, batch_size, **kwargs):
-        _, num_classes, X_train, y_train, X_valid, y_valid = load_mnist_shard(
-            shard_num=int(data_path), **kwargs
-        )
+        try:
+            _, num_classes, X_train, y_train, X_valid, y_valid = load_mnist_shard(
+                shard_num=int(data_path), **kwargs
+            )
+        except ValueError:
+            logger.error("Please pass the shard number (integer) for the collaborator using data path flag.")
+            return
```
This block will raise `ValueError` on any downstream error that may occur within `try` block. Our intention is only to check the typecast.

Recommended way:
```python
try:
  shard_num = int(data_path)
except:
  raise ValueError(...)  # auto-exits.

_, num_classes, X_train, y_train, X_valid, y_valid = load_mnist_shard(
  shard_num=shard_num, **kwargs)
```

theakshaypant replied:
Added the change in 3035e0ed71cfc12bff6f13ad64262a43071f799d.

MasterSkepticista reviewed (changes requested):
Change the scope of `try...except` block

theakshaypant committed:
fix(openfl-workspaces): added type check for data path as shard num (int) before loading dataset

Signed-off-by: Pant, Akshay <akshay.akshaypant@gmail.com>

MasterSkepticista commented:
@theakshaypant Please rebase your branch with `develop`. PKI test needs to pass.

MasterSkepticista commented on openfl-workspace/keras_cnn_mnist/src/tfmnist_inmemory.py:
```diff
@@ -26,6 +26,10 @@ def __init__(self, data_path, batch_size, **kwargs):
+        try:
+            int(data_path)
+        except:
+            raise ValueError("Pass shard number using data path flag as an int.")
```
General: `Exception`s, warnings and assertions should inform the user precisely what went wrong. In this case, one could rephrase it as 

```python
ValueError("Expected `%s` to be representable as `int`.", data_path)
```

More [examples](https://google.github.io/styleguide/pyguide.html#3102-error-messages)

theakshaypant replied:
Updated in c6e35bf304c47b1ad02e73ad4f0a294bbb53721c.

MasterSkepticista reviewed (approved):
LGTM

theakshaypant committed:
fix(openfl-workspaces): changed error message for ValueError

Signed-off-by: Pant, Akshay <akshay.akshaypant@gmail.com>

theakshaypant committed:
Merge branch 'develop' of https://github.com/theakshaypant/openfl into akshay/fix-read_shard_from_default_path

theakshaypant committed:
restructure: linter changes

Signed-off-by: Pant, Akshay <akshay.akshaypant@gmail.com>

teoparvanov commented on docs/about/features_index/taskrunner.rst:
```diff
@@ -382,7 +382,7 @@ Importing the Workspace
     .. code-block:: console
 
-       $ fx collaborator create -n {COL_LABEL} -d {DATA_PATH:optional}
+       $ fx collaborator create -n {COL_LABEL} -d {DATA_PATH}
```
In general, the entire `-d {DATA_PATH}` should be an optional parameter, to cover cases where the data path is already set in `data.yaml` as part of the FL plan creation. Not sure what the `:optional` refers to here though - is it just for the DATA_PATH? If so, then your suggestion is correct. If not, we should leave the documentation as is.

theakshaypant replied:
In some examples (the ones modified by this PR), data path represents the shard number to be used for that collaborators. When we use the default value [default_data_path = f"data/{collaborator_name}"](https://github.com/securefederatedai/openfl/blob/develop/openfl/interface/collaborator.py#L140), an error is encountered when starting the collaborator is started. 
```
Data = {'c1': 'data/c1', 'c2': 'data/c2'}
           INFO     🧿 Starting a Collaborator Service.                      collaborator.py:99
           INFO     Building `src.dataloader.PyTorchMNISTInMemory` Module.   plan.py:228
EXCEPTION : invalid literal for int() with base 10: 'data/c1'
Traceback (most recent call last):
  File "/root/openfl-maestro/bin/fx", line 8, in <module>
    sys.exit(entry())
  File "/root/openfl-maestro/lib/python3.8/site-packages/openfl/interface/cli
.py", line 328, in entry
    error_handler(e)
  File "/root/openfl-maestro/lib/python3.8/site-packages/openfl/interface/cli
.py", line 246, in error_handler
    raise error
  File "/root/openfl-maestro/lib/python3.8/site-packages/openfl/interface/cli
.py", line 326, in entry
    cli(max_content_width=120)
  File "/root/openfl-maestro/lib/python3.8/site-packages/click/core.py", line
1157, in __call__
    return self.main(*args, **kwargs)
  File "/root/openfl-maestro/lib/python3.8/site-packages/click/core.py", line
1078, in main
    rv = self.invoke(ctx)
  File "/root/openfl-maestro/lib/python3.8/site-packages/click/core.py", line
1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/root/openfl-maestro/lib/python3.8/site-packages/click/core.py", line
1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/root/openfl-maestro/lib/python3.8/site-packages/click/core.py", line
1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/root/openfl-maestro/lib/python3.8/site-packages/click/core.py", line
783, in invoke
    return __callback(*args, **kwargs)
  File "/root/openfl-maestro/lib/python3.8/site-packages/openfl/interface/col
laborator.py", line 101, in start_
    plan.get_collaborator(collaborator_name).run()
  File "/root/openfl-maestro/lib/python3.8/site-packages/openfl/federated/pla
n/plan.py", line 601, in get_collaborator
    data_loader = self.get_data_loader(collaborator_name)
  File "/root/openfl-maestro/lib/python3.8/site-packages/openfl/federated/pla
n/plan.py", line 453, in get_data_loader
    self.loader_ = Plan.build(**defaults)
  File "/root/openfl-maestro/lib/python3.8/site-packages/openfl/federated/pla
n/plan.py", line 235, in build
    instance = getattr(module, class_name)(**settings)
  File "/home/akshay/akshay/src/dataloader.py", line 30, in __init__
    shard_num=int(data_path), **kwargs
ValueError: invalid literal for int() with base 10: 'data/c1'
```

Hence at the time of creating this PR, I suggested modifying the documentation to remove the "optional" for DATA_PATH. 
We can keep the documentation as is with a more descriptive error message in the example as you suggested in another comment.

theakshaypant replied:
Reverted change in aae8185f8ab1ba1ce58cd4ff72bc9931b926bb73.

teoparvanov commented on openfl-workspace/torch_cnn_histology/src/dataloader.py:
```diff
@@ -37,9 +37,14 @@ def __init__(self, data_path, batch_size, **kwargs):
+        try:
+            int(data_path)
+        except:
+            raise ValueError("Expected `%s` to be representable as `int`.", data_path)
```
In the error message, specify that this is for the shard number for the given collaborator.

theakshaypant replied:
Addressed in ca90679af5ca4788e1bc8886dda550dd34f57a0d.

teoparvanov commented on openfl-workspace/torch_cnn_histology_gramine_ready/src/pthistology_inmemory.py:
```diff
@@ -21,8 +21,14 @@ def __init__(self, data_path, batch_size, **kwargs):
+        try:
+            int(data_path)
+        except:
+            raise ValueError("Expected `%s` to be representable as `int`.", data_path)
```
Same comment here, and for the other `ValueError` instances throughout the PR

theakshaypant replied:
Addressed in ca90679af5ca4788e1bc8886dda550dd34f57a0d.

teoparvanov reviewed (approved):
LGTM, with a couple of comments

teoparvanov commented on openfl-workspace/torch_unet_kvasir/src/data_loader.py:
```diff
@@ -121,6 +122,11 @@ def __init__(self, data_path, batch_size, **kwargs):
         """
         super().__init__(batch_size, **kwargs)
 
+        try:
+            int(data_path)
+        except:
+            raise ValueError("Expected `%s` to be representable as `int`.", data_path)
+
```
By the way, I suppose you've re-tested all the workspaces modified in this PR, right?
If so, please add a "Testing done" section to the PR description.

theakshaypant replied:
Had tested the change. It is available in the original description of the PR which has since been modified.

theakshaypant committed:
Merge branch 'develop' of https://github.com/theakshaypant/openfl into akshay/fix-read_shard_from_default_path

theakshaypant committed:
fix(task runner examples): add a more descriptive error message

Signed-off-by: Pant, Akshay <akshay.akshaypant@gmail.com>

theakshaypant committed:
fix(task runner docs): revert to DATA_PATH being optional

Signed-off-by: Pant, Akshay <akshay.akshaypant@gmail.com>
